architecture-docs/adr-003.md

```diff
@@ -0,0 +1,144 @@
+# ADR 003: Migrate to jellyfish merkle trie
+
+## Changelog
+* 31-03-2020: Initial Draft
+
+## Context
+
+Need a merkle trie library to store staking state, current one has several disadvantages:
+
+- Written and maintained by single person.
+- Custom data structure and algorithm.
+- Not enough documentation.
+
+## Decision
+
+### jellyfish merkle trie
+
+https://github.com/libra/libra/blob/master/storage/jellyfish-merkle/src/lib.rs
+
+#### Advantages
+
+- Review from Facebook security engineers
+- Small code base, well documented.
+- Well known data structure (simplified version of ethereum MPT).
+- Easy to integrate.
+- Optimized for pbft: Use u64 version number instead of hash.
+- License is apache 2.0.
+- Support inclusion proof and range proof.
+
+#### Disadvantages
+
+- Tied with a lot of custom Libra crates, need to extract an independent library.
+
+  Initial work has done here: https://github.com/yihuang/libra
+
+- No extension node compared to ethereum MPT, deeper tree in theory, a trade off made in favor of simplicity.
+
+### Integration
```

Comment:
how will this be done? one giant PR or PR for adding the crate and then PR for implementing readers etc.? or ?

Reply:
I estimates that one PR should be good enough, won't be too big.

Reply:
should be mentioned.

Also "ChainState" integrations -- will/can this support historical querying? 
My assumption is that this can support historical querying, but instead of using the root hash, one can use the "u64" version number?

Reply:
> should be mentioned.
> 
> Also "ChainState" integrations -- will/can this support historical querying?
> My assumption is that this can support historical querying, but instead of using the root hash, one can use the "u64" version number?

Yes, use the block height as the version number.

Reply:
so I guess that's a positive impact that there's no need to store that `APP_HASH -> block height` mapping for historical querying?

Reply:
> so I guess that's a positive impact that there's no need to store that `APP_HASH -> block height` mapping for historical querying?

Yes, I'll add that.

Reply:
> should be mentioned.
Done.
Added: "Implementation plan", "Staled nodes cleanup strategy", "Chain node state changes",

Reply:
Changed the implementation plans to use two PRs.